onnxscript/rewriter/pattern.py

```diff
@@ -1029,37 +1033,118 @@ def _match_node_output(self, pattern_value: NodeOutputPattern, value: ir.Value)
+                "Internal Error: SimplePatternMatcher should not be used for patterns with multiple output nodes."
             )
-        if pattern._output_node is None:
+
+        if not self._match_node(pattern.output_node, node):
+            return match
+
+        output_values = self._get_output_values()
+        if output_values is None:
+            return match
+        if not _valid_to_replace(match.nodes, output_values):
+            return match.fail("Matched nodes have other uses preventing replacement.")
+
+        if len(node.outputs) != pattern.num_outputs:
             return match.fail(
-                "Internal Error: SimplePatternMatcher should not be used for patterns with multiple output nodes."
+                f"Number of node outputs mismatch: expected {pattern.num_outputs}, got {len(node.outputs)}."
             )
 
-        if self._match_node(pattern._output_node, node):
-            if not _valid_to_replace(match.nodes):
-                return match.fail("Matched nodes have other uses preventing replacement.")
+        match.outputs.extend(output_values)
+        return match
+
+    def _multi_match(self, candidate: dict[NodePattern, ir.Node]) -> MatchResult:
+        match = self._match
+        for pattern_node, node in candidate.items():
+            if not self._match_node(pattern_node, node):
+                return match
+        output_values = self._get_output_values()
+        if output_values is None:
+            return match
 
-        match.outputs.extend(node.outputs)
+        if not _valid_to_replace(match.nodes, output_values):
+            return match.fail("Matched nodes have other uses preventing replacement.")
+
+        match.outputs.extend(output_values)
         return match
 
+    def match(
+        self,
+        model: ir.Model,
+        graph_or_function: ir.Graph | ir.Function,
+        node: ir.Node,
+        verbose: int = 0,
+    ) -> MatchResult:
+        if self.pattern.has_single_output_node:
+            self._init_match(verbose)
+            return self._match_single_output_node(model, graph_or_function, node)
+        else:
+            pattern_output_nodes = self.pattern.output_nodes
+            op_to_nodes: dict[tuple[str, str, str], list[ir.Node]] = {}
+            for n in graph_or_function:
+                op_to_nodes.setdefault(n.op_identifier(), []).append(n)
+            all_nodes = iter(graph_or_function)
+
+            def get_nodes(pattern_node):
+                id = pattern_node.op_identifier()
+                if id is None:
+                    return all_nodes
+                return op_to_nodes.get(id, [])
+
+            candidates = [iter([node])] + [get_nodes(pn) for pn in pattern_output_nodes[1:]]
+            match = None
+            for combination in itertools.product(*candidates):
+                candidate = dict(zip(pattern_output_nodes, combination))
```

Comment:
I would add a comment to describe the algo/logic of this loop (what `product` does etc.)